crates/viewer/re_space_view_graph/src/layout/provider.rs

```diff
@@ -0,0 +1,149 @@
+        let simulation = fj::SimulationBuilder::default()
+            .with_alpha_decay(0.01) // TODO(grtlr): slows down the simulation for demo
+            .build(all_nodes)
+            .add_force(
+                "link",
+                fj::Link::new(all_edges).distance(50.0).iterations(2),
+            )
+            .add_force("charge", fj::ManyBody::new())
+            // TODO(grtlr): This is a small stop-gap until we have blueprints to prevent nodes from flying away.
+            .add_force("x", fj::PositionX::new().strength(0.01))
+            .add_force("y", fj::PositionY::new().strength(0.01));
+
+        Self {
+            simulation,
+            node_index,
+            edges: all_edges_iter.copied().collect(),
+        }
+    }
+
+    pub fn init(&self, request: &LayoutRequest) -> Layout {
+        let positions = self.simulation.positions().collect::<Vec<_>>();
+        let mut extents = ahash::HashMap::default();
+
+        for graph in request.graphs.values() {
+            for (id, node) in &graph.nodes {
+                let i = self.node_index[id];
+                let [x, y] = positions[i];
+                let pos = Pos2::new(x as f32, y as f32);
+                extents.insert(*id, Rect::from_center_size(pos, node.size));
+            }
+        }
+
+        Layout {
+            nodes: extents,
+            // Without any real node positions, we probably don't want to draw edges either.
+            edges: ahash::HashMap::default(),
+        }
+    }
+
+    /// Returns `true` if finished.
+    pub fn tick(&mut self, layout: &mut Layout) -> bool {
+        self.simulation.tick(1);
+
+        let positions = self.simulation.positions().collect::<Vec<_>>();
+
+        for (node, extent) in &mut layout.nodes {
+            let i = self.node_index[node];
+            let [x, y] = positions[i];
+            let pos = Pos2::new(x as f32, y as f32);
+            extent.set_center(pos);
+        }
+
+        for (from, to) in &self.edges {
+            layout.edges.insert(
+                (*from, *to),
+                line_segment(layout.nodes[from], layout.nodes[to]),
+            );
+        }
+
+        self.simulation.finished()
+    }
+}
+
+/// Helper function to calculate the line segment between two rectangles.
+fn line_segment(source: Rect, target: Rect) -> [Pos2; 2] {
+    let source_center = source.center();
+    let target_center = target.center();
+
+    // Calculate direction vector from source to target
+    let direction = (target_center - source_center).normalized();
+
+    // Find the border points on both rectangles
+    let source_point = find_border_point(source, -direction); // Reverse direction for target
+    let target_point = find_border_point(target, direction);
+
+    [source_point, target_point]
+}
+
+/// Helper function to find the point where the line intersects the border of a rectangle
+fn find_border_point(rect: Rect, direction: Vec2) -> Pos2 {
```

Comment:
It would be nice with a unit-test for this.

Also: what is the line defined by the direction?

Btw, this is all pretty similar to https://docs.rs/emath/latest/src/emath/rect.rs.html#652-674 - maybe we can lift in the function into `emath`

Reply:
Done! If you like the code as it is, I'm happy to uptream to `emath`! I tried to give it a name similar to `Rect::intersects_ray`.